src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1085,6 +1085,36 @@ private void WriteHeaders(HttpRequestMessage request)
             }
         }
 
+        private HttpRequestException GetShutdownException()
+        {
+            Debug.Assert(Monitor.IsEntered(SyncObject));
+
+            // Throw a retryable exception that will allow this unprocessed request to be processed on a new connection.
+            // In rare cases, such as receiving GOAWAY immediately after connection establishment, we will not 
+            // actually retry the request, so we must give a useful exception here for these cases.
+
+            Exception innerException;
+            if (_abortException != null)
+            {
+                innerException = _abortException;
+            }
+            else if (_lastStreamId != -1)
+            {
+                // We must have received a GOAWAY.
+                innerException = new IOException(SR.net_http_server_shutdown);
+            }
+            else
+            {
+                // We must either be disposed or out of stream IDs.
+                // Note that in this case, the exception should never be visible to the user (it should be retried).
```

Comment:
Aren't we going to retry for all of these?  Is this comment saying that whereas the if and the else if may end up propagating to the user in the rare race condition mentioned, this case will never be subject to that rare race condition?  If so, why do we bother creating an ODE exception at all, rather than just letting innerException be null?

Reply:
> Aren't we going to retry for all of these?

Yes

> Is this comment saying that whereas the if and the else if may end up propagating to the user in the rare race condition mentioned, this case will never be subject to that rare race condition?

Yes. The other cases are caused by server issues or network failure. This case is caused by us, deciding not to reuse the connection. As such, this should never be the first request on the connection, and so should always be retried.

> If so, why do we bother creating an ODE exception at all, rather than just letting innerException be null?

Mainly just so that if it ever does show up in the wild (and it shouldn't), it will be obvious where this exception came from.

Reply:
Ok